src/main/java/com/scalar/cassy/transferer/RemoteFileSystemFileUploader.java

```diff
@@ -0,0 +1,130 @@
+package com.scalar.cassy.transferer;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.palantir.giraffe.file.MoreFiles;
+import com.scalar.cassy.config.BackupConfig;
+import com.scalar.cassy.exception.FileTransferException;
+import com.scalar.cassy.util.RemoteFileSystemConnection;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RemoteFileSystemFileUploader extends RemoteFileSystemTransfererBase
+    implements FileUploader {
+  private static final Logger logger = LoggerFactory.getLogger(RemoteFileSystemFileUploader.class);
+  private static final int NB_PARALLEL_UPLOAD_THREADS = 1;
+
+  @Inject
+  public RemoteFileSystemFileUploader(RemoteFileSystemConnection hostConnection) {
+    super(hostConnection);
+  }
+
+  @Override
+  public Future<Void> upload(Path file, String key, String storageBaseUri) {
+    String remoteFileSystemStoragePath = URI.create(storageBaseUri).getPath();
+    Path targetFilePath =
+        hostConnection.getHostControlSystem().getPath(remoteFileSystemStoragePath, key);
+    File targetFile = new File(targetFilePath.toString());
+    if (targetFile.exists() && targetFile.length() == file.toFile().length()) {
+      logger.info(file + " has been already uploaded.");
+      return CompletableFuture.completedFuture(null);
+    }
+
+    logger.info("Uploading " + file);
+    try {
+      createDirectories(targetFilePath.getParent());
+      copyFile(file, targetFilePath);
+      return CompletableFuture.completedFuture(null);
+    } catch (IOException e) {
+      throw new FileTransferException(e);
+    }
+  }
+
+  @Override
+  public void upload(List<Path> files, BackupConfig config) {
```

Comment:
The uploader now copy file by file using giraffe `MoreFiles.copy`. 
The uploads tasks are run by the executorService which is configured with one thread at the moment.